lib/kubernetes-deploy/discoverable_resource.rb

```diff
@@ -0,0 +1,175 @@
+    end
+
+    def self.discover_tpr(client)
+      return unless client.respond_to? :get_third_party_resources
+      resources = client.get_third_party_resources
+      resources.each do |res|
+        type, _, group = res.metadata.name.partition('.')
+        # TPR API supports multiple versions for a single resource :(
+        res.versions.each do |version|
+          discovered(group: group,
+                     type: type,
+                     version: version.name,
+                     annotations: res.metadata.annotations)
+        end
+      end
+    end
+
+    def self.discover_crd(client)
+      return unless client.respond_to? :get_custom_resource_definitions
+      resources = client.get_custom_resource_definitions
+      resources.each do |res|
+        discovered(group: res.spec.group,
+                   type: res.spec.names.kind,
+                   version: res.spec.version,
+                   annotations: res.metadata.annotations)
+      end
+    end
+
+    def self.discovered(group:, type:, version:, annotations:)
+      resource_class = Class.new(self) do
+        type = type.capitalize # TPRs are inconsistent about capitalization
+        @group = group
+        @type = type
+        @version = version
+        @prunable = DiscoverableResource.parse_bool(annotations[PRUNABLE_ANNOTATION])
+        @predeploy = DiscoverableResource.parse_bool(annotations[PREDEPLOY_ANNOTATION])
+        @timeout = DiscoverableResource.parse_timeout(type, annotations[TIMEOUT_ANNOTATION])
+
+        status_field = annotations[STATUS_FIELD_ANNOTATION]
+        success_status = annotations[STATUS_SUCCESS_ANNOTATION]
+
+        define_method 'deploy_succeeded?' do
+          getter = "get_#{type.downcase}"
+          @client ||= DiscoverableResource.kubeclient(context: @context, resource_class: self.class)
+          raw_json = @client.send(getter, @name, @namespace, as: :raw)
+          query_path = JsonPath.new(status_field)
+          current_status = query_path.first(raw_json)
+          current_status == success_status
+        end if status_field && success_status
+      end
+
+      add_resource(resource_class)
+    end
+
+    def self.add_resource(resource_class)
+      group = resource_class.group
+      type = resource_class.type
+      version = resource_class.version
+      @resources ||= {}
+      @resources[group] ||= {}
+      @resources[group][type] ||= {}
+      @resources[group][type][version] = resource_class
+    end
+
+    def self.parse_bool(value)
+      value.to_s == "true" # value could be nil
+    end
+
+    def self.parse_timeout(type, timeout)
+      ActiveSupport::Duration.parse(timeout) if timeout
+    rescue ActiveSupport::Duration::ISO8601Parser::ParsingError
+      raise FatalDeploymentError,
+        "Resource #{type} specified invalid timeout value '#{timeout}', must use ISO8601 duration."
+    end
+
+    def self.kubeclient(context:, resource_class:)
+      _build_kubeclient(
```

Comment:
This value cannot be memoized because we must refresh what CRD/TPRs are available on each invocation (even if it's during a single lifespan of a`k8s-deploy` process, for example if using this code as a library instead of a standalone executable during test).

Reply:
Is that because we might have deployed CRDs? If we want to make that work, we should predeploy them before doing the discovery (or maybe just accept that this does not work for v1, since that'd be really complicated). Doing discovery every time we need to use the client sounds expensive. Alternatively, could we be using instances of this class and discarding them on every `run`?